### ISSUE TITLE
_mm_cmpXstrc() not coalesced with a subsequent _mm_cmpXstrm() or  _mm_cmpXstri()

### ISSUE THREAD

**Issue by dtrebbien** [rust-lang/rust#53232]
If an \_mm_cmpestrc() intrinsic is used to check whether a match was found, and then an \_mm_cmpestrm() intrinsic is used to obtain the mask __m128i, the assembly that is generated by rustc contains a (V)PCMPESTRI instruction and a (V)PCMPESTRM instruction. Ideally, the assembly would have only one (V)PCMPESTRM instruction.

The situation is the same for the following combinations:

 * \_mm_cmpestrc() followed by an \_mm_cmpestri() generates two (V)PCMPESTRI instructions
 * \_mm_cmpistrc() followed by an \_mm_cmpistrm() generates a (V)PCMPISTRI instruction and a (V)PCMPISTRM instruction
 * \_mm_cmpistrc() followed by an \_mm_cmpistri() generates two (V)PCMPISTRI instructions

Over at shepmaster/jetscii#24, I prepared a Rust test case and a functionally equivalent C++ test case: https://github.com/shepmaster/jetscii/pull/24#issuecomment-411861393

I found that GCC 8.2.0 is able to coalesce the (V)PCMPxSTRx instructions. However, LLVM 6.0.1 is not. This appears to be a known issue in LLVM: https://bugs.llvm.org/show_bug.cgi?id=37246

This issue particularly impacts efficient use of the "mask" variants (explicit and implicit length); the "index" variants can simply check whether the index is 16, which indicates no match.

I am posting this issue here so that people can more easily find this information, but I am not sure what stdsimd can do. Perhaps, though, one possibility is to provide "mask" variant combination intrinsics, say cmpestrm() and cmpistrm(), which return the CFlag *and* mask XMM register.

**Comment by gnzlbg** [comment]
> but I am not sure what stdsimd can do

I've crosslinked this in the LLVM issue. It seems that Chandler wanted this to be fixed, I'll ping him on IRC to see if either he has time, or if he can mentor somebody that has time. We are currently using LLVM-trunk in rustc, so if this were to be fixed in LLVM then it would be fairly easy to upgrade to it. 

However, I don't think there are any workarounds that are feasible for `stdsimd`. I've asked in the LLVM bug, maybe somebody there knows what we could do. 

In the mean time, if you _really_ need this right now, and can live with nightly, inline `asm!` might currently be your only option :/

**Comment by dtrebbien** [comment]
I was experimenting with my test case, and I came up with:

```rust
// rustup run nightly rustc -C opt-level=3 -C target-cpu=native -o main_exp_rustc main_exp.rs && ./main_exp_rustc
// rustup run nightly rustc -C opt-level=3 -C target-cpu=native --emit asm -o main_exp_rustc.s main_exp.rs

#![feature(asm)]

use std::arch::x86_64::*;

union XmmTransmute {
    epi8: [i8; 16],
    xmm: __m128i,
}

/// **Rust extension:** Compare packed strings in `a` and `b` with lengths `la`
/// and `lb` using the control byte in `imm8`, and return a tuple of:
/// - the resulting mask
/// - the CF (1 if the resulting mask was non-zero; otherwise, 0)
/// - the ZF (1 if -16 < `lb` < 16; otherwise, 0)
/// - the SF (1 if -16 < `la` < 16; otherwise, 0)
/// - the OF (the least significant bit of the resulting mask)
#[inline(always)]
fn cmpestrm_and_flags(a: __m128i, la: i64, b: __m128i, lb: i64, imm8: u8) -> (__m128i, i32, i32, i32, i32) {
    let mask: __m128i;
    let status_flags: u16;
    // GCC supports "flag output constraints" (e.g. =@ccc for the CF):
    // https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#index-asm-flag-output-operands
    // This would be great to use, but it appears that LLVM does not have a
    // similar feature:
    // http://llvm.org/docs/LangRef.html#supported-constraint-code-list
    unsafe {
        asm!("pcmpestrm $6, $4, $2
              ${:comment} lahf stores the CF, PF, AF, ZF, and SF into %ah, but not the OF.
              ${:comment} We'll place OF at bit 0 of %ax by conditionally moving.
              mov $$0, %ax
              mov $$1, %dx
              cmovo %dx, %ax
              lahf"
             : "={xmm0}"(mask), "={ax}"(status_flags)
             : "x"(a), "{rax}"(la), "x"(b), "{rdx}"(lb), "i"(imm8)
             : "flags");
    }
    let cf = ((status_flags >> 8) & 1_u16) as i32;
    let zf = ((status_flags >> 14) & 1_u16) as i32;
    let sf = ((status_flags >> 15) & 1_u16) as i32;
    let of = (status_flags & 1_u16) as i32;
    (mask, cf, zf, sf, of)
}

fn main() {
    let haystack = "the quick brown fox jumps over a lazy dog\u{0}";

    unsafe {
        let haystack = haystack.get_unchecked(..16);
        println!("haystack = '{}'", haystack);

        let needle_xmm = XmmTransmute { epi8: ['a' as i8, 'e' as i8, 'i' as i8, 'o' as i8, 'u' as i8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] }.xmm;
        let haystack_xmm = _mm_loadu_si128(haystack.as_ptr() as *const __m128i);

        println!("\nTrying PCMPESTRM...");
        let (mask_xmm, cf, zf, sf, of) = cmpestrm_and_flags(needle_xmm, 5, haystack_xmm, 16, 0);
        println!("cf = {}, zf = {}, sf = {}, of = {}", cf, zf, sf, of);
        if cf != 0 {
            let mask = _mm_extract_epi32(mask_xmm, 0) as u16;
            println!("PCMPESTRM found {} vowel(s). mask = 0b{:016b}", mask.count_ones(), mask);
        } else {
            println!("No vowel was found by PCMPESTRM.");
        }
    }
}
```

As I mentioned in a comment, it would be really nice if LLVM supports "flag output constraints" like GCC, but I wasn't seeing anything like this in the LLVM lang ref. Does anyone know if LLVM actually does support this feature?

Also, one issue: I am not sure why input constraints "{rax}" and "{rdx}" are not being followed; the assembly output shows that `la` and `lb` are being moved into %eax and %edx:

```asm
	movl	$5, %eax
	movl	$16, %edx
	vmovaps	-176(%rbp), %xmm0
	vmovaps	-208(%rbp), %xmm1
	## InlineAsm Start
	pcmpestrm	$0, %xmm1, %xmm0
	## lahf stores the CF, PF, AF, ZF, and SF into %ah, but not the OF.
	## We'll place OF at bit 0 of %ax by conditionally moving.
	movw	$0, %ax
	movw	$1, %dx
	cmovow	%dx, %ax
	lahf
	## InlineAsm End
```

**Comment by gnzlbg** [comment]
@dtrebbien no idea. You might want to give the LLVM bug tracker https://bugs.llvm.org a try, and see if you can find an LLVM bug open about either of these things. If not, we should try to minimize the two cases, and fill the two bugs. If you want to give that a try you could:

* produce a _minimal_ rust example in https://rust.godbolt.org
* use the following options to generate LLVM-IR
    -C opt-level=N (0, 3?) -C lto=fat -C panic=abort -C debuginfo=0 --emit=llvm-ir / --emit=asm

Once you have an example with LLVM-IR close to what you want, you can try to modify it and see if it can do what you wanted (maybe it is just the Rust front-end that doesn't support these). The easiest way would be to use: https://gcc.godbolt.org and choose as language LLVM-IR. There are two tools there: `opt` (optimizer) and `llc` (machine code generation). You can use llc to generate assembly. 

If Rust generates correct LLVM-IR, but llc produces suboptimal / incorrect machine code, this is definetely an LLVM bug. If the LLVM-IR doesn't allow you to specify some constraints, you can fill an LLVM showing which assembly you get without the constraints, and which assembly you wanted, and use that to fill another LLVM bug.

You can share all the godbolt links, so feel free to include them in the bug reports, but it is also good practice to copy the input and output to the bug reports in case godbolt gets hit by a bus.

**Comment by dtrebbien** [comment]
@gnzlbg Thank you for this information! I will try to prepare a minimal Rust example.

**Comment by dtrebbien** [comment]
> Also, one issue: I am not sure why input constraints "{rax}" and "{rdx}" are not being followed

Of course!  When movl is used to move a 32-bit value into a 64-bit register, the value is zero-extended: https://stackoverflow.com/questions/11177137/why-do-x86-64-instructions-on-32-bit-registers-zero-the-upper-part-of-the-full-6

So, the "{rax}" and "{rdx}" input constraints *are* being followed.

Just to double check, I examined the IR and saw that rustc wrote out the constraints verbatim.
